### PR TITLE
Compute the viscosity as a function of the temperature and the shear stress

### PR DESCRIPTION
**Pull request by AmelieLaurens**
Organize in classes and modules the code :
- create a file viscosity.yaml to put the required values
- create several classes : Deck : get the value in viscosity.yaml
                                           Polymer : stock the values of deck in variables that we will be reuse
                                           Model : contain the equation to predict the viscosity
                                           Graph : draw the graphic and save it in the folder Graphics
- main_viscosity.py : the only file which needs to be run

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+pyyaml
+numpy
+matplotlib
+math
```
Comment (jaklengaigne) [inline requirements.txt]: It seems that the math package cannot be installed. 
I am not even sure you actually need it, i was able to run the code on fresh conda environment without it. @ilyasst Can you confirm?
image

File: rotary_jet_spinning/datas/data.py [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+        self.final_radius_density(deck, polymer, machine, model)
+        self.final_radius_surface_tension(deck, polymer, machine, model)
+        self.final_radius_viscosity(deck, polymer, machine, model)
+        self.radius_x(deck, polymer, machine, model)
+        
+
+
+    def final_radius_orifice_radius(self, deck, polymer, machine, model):
+        
+        discretisation = int(deck.doc['Discretisation'])
```
Comment (jaklengaigne) [inline rotary_jet_spinning/datas/data.py]: Is there a reason why discretisation is not define as a global variable in the init of the class?
You are using it a lot in the different functions, it seems ineffective to have it redefined each time.

**Comment by ilyasst** [comment]
Bump @jaklengaigne :)

File: rotary_jet_spinning/datas/data.py [hunk 3]
```diff
@@ -0,0 +1,253 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+"""
+Source : A simple model for nanofiber formation by rotary jet-spinning
```
Comment (jaklengaigne) [inline rotary_jet_spinning/datas/data.py:4]: please at the full bibliographic details here.

What do you think of renaming the class to signal that it is only for mellado's model? When we will add new models or model correction we will be able to keep track.
@ilyasst , can you weight in for such a nomenclature? How do you think we should organize the different models and their corresponding codes?

File: rotary_jet_spinning/datas/data.py (hunk 2 above)
Comment (jaklengaigne) [reply rotary_jet_spinning/datas/data.py]: Same question for nu?
The kinetic velocity is used in every function. Defining it every time seems counterproductive.

File: README.md [hunk 4]
```diff
@@ -2,6 +2,8 @@
```
Comment (jaklengaigne) [inline README.md:2]: I am going to do a PR to adjust the README.md.

File: rotary_jet_spinning/datas/data.py [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+        Final_radius = []
+        for k in range(discretisation):
+            Final_radius.append(model.final_radius(orifice_radius[k], initial_velocity[k],
+                                            nu, machine.collector_radius, machine.omega))
+
+        return [orifice_radius, Final_radius]
+        
+
+    
+    def final_radius_omega(self, deck, polymer, machine, model):
```
Comment (jaklengaigne) [inline rotary_jet_spinning/datas/data.py]: omega is the angular velocity. In the deck.yaml it is named as such. 
I think that for consistencies, the name should be keep the same. If possible replace omega by angular_velocity where applicable. It won't be pretty in some places, but it will maintain consistency and help readability i think.

File: viscosity/models/modelling.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+"""
+Reference :
```
Comment (jaklengaigne) [inline viscosity/models/modelling.py]: Excellent! Do the same for Mellado's model.

**Review by jaklengaigne — COMMENTED** [review]
The code runs without trouble. Just double check for the math library. Graphs render nicely and are quite clear. I believe the code is nearing its final form and we should soon be able to add some experimental data as examples. I have to main comments regarding the documentation of the code and a maybe more high level discussion on how to implement more user defined variable ranges. See below.


**Documentation:**
I think that the code needs a bit more descriptive comments on what it does. I am thinking especially about the different functions defined in data.py, but that is also the case for the other. 
I am not sure exactly how much we should details, but at least a single sentence at the beginning of each function to state its purpose would be necessary.
This is not regarding the README, as mention in comment, I will be submitting a PR soon about that.

**User input control**
I think we may have to discuss how the code is made to compute the different results. There is a mix of input variables and in-code parameter generation. 
Let me try to explain, the user has to input the machine and polymer parameters (in the deck file), these parameters form a set of unique properties (one single viscosity, one single reservoir size, etc). Then in the Data class, the code itself choose to vary each parameter in a predetermined range to show its effect on the final fiber diameter, keeping the rest of the input fixed to the user-values. For now there is no way for the user to change these ranges without messing in the code itself.
How can we implement these range exploration in the deck input file? Should we even do that?
I think we need a call to discuss that.

File: rotary_jet_spinning/datas/data.py (hunk 3 above)
Comment (ilyasst) [reply rotary_jet_spinning/datas/data.py:4]: We could include a .bibfile in the repo (it would be easy as we already have the library on Zotero, we just need to export it as a bib file) and use the unique identifiers for each publication in the python code (such as [mellado2011]). It makes it easier to explore the code and "insert citation" within it. If someone wants the details of the paper, they can just open the bibfile in the repo and access them.

File: requirements.txt (hunk 1 above)
Comment (jaklengaigne) [reply requirements.txt]: @ilyasst bump?

Comment (ilyasst) [reply requirements.txt]: Sorry missed this one, `math` is part of the Python standard library https://docs.python.org/3/library/math.html so it is always included with Python. It is impossible to install.

**Comment by ilyasst** [comment]
@jaklengaigne can you take care of the review for the new version of this PR since you worked with Amelie on the last one ?

File: main.py [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+
+data = Data(deck, polymer, machine, model, features)
+
+organization = Organization(data, deck, machine, polymer, model, features)
+
+organized_data = organization.organize_data(data, deck, machine, polymer, model, features)
+
+graph = PointGraph(organized_data)
+
+# comparison = Comparison(deck, polymer, machine, model)
```
Comment (jaklengaigne) [inline main.py:27]: Just to make sure, the comparison has not yet been implemented (or is not part of this PR)?

Comment (AmelieLaurens) [reply main.py:27]: It is not part of this PR. I am waiting for this one to be closed before doing a PR for the comparison with experiment part.

File: rotary_jet_spinning/models/modelling.py [hunk 8]
```diff
@@ -2,24 +2,24 @@
 # -*- coding: utf-8 -*-
 
 from math import *
```
Comment (jaklengaigne) [inline rotary_jet_spinning/models/modelling.py]: why do you import all math?
if you are using only a few math function you can import only these. e.g. if you use only sqrt() you can do:
`from math import sqrt`

**Review by jaklengaigne — APPROVED** [review]
The code ran without errors in a fresh conda environment. I am going to approve and merge!
Only one real small comment about the math library. You can do it in the next PR. 

I am going to add an issue to improve separately the readme.

File: rotary_jet_spinning/models/modelling.py [hunk 9]
```diff
@@ -34,7 +34,8 @@ def critical_rotational_velocity_threshold(self, surface_tension, orifice_radius
         
     def Initial_velocity(self, omega_th, s0):
```
Comment (jaklengaigne) [inline rotary_jet_spinning/models/modelling.py]: small error in the name. It should not be a capital first later.

Comment (AmelieLaurens) [reply rotary_jet_spinning/models/modelling.py]: I put a capital to the name of the function because I already used initial_velocity as a variable name. Should I change the name of the function by initial_velocity_fct for example or can I keep the capital letter Initial_velocity ?

Comment (jaklengaigne) [reply rotary_jet_spinning/models/modelling.py]: Ok that is a conflict we shoud resolve. Are both of them referring to the same thing? If not, one of them may be not well named.
Is the variable initialized using the function? If so, what about changing the variable name to v_ini

Comment (AmelieLaurens) [reply rotary_jet_spinning/models/modelling.py]: yes there are referring to the same thing, the variable is initialized with the function yes. ok I changed the variable name to init_velocity

**Comment by ilyasst** [comment]
Should this one be closed or is it still active ?

**Comment by jaklengaigne** [comment]
I believe that it is closed since I merged it.
It was a revert action to correct my wrong merging.
